peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+
+* annotating the attribute with :data:`typing.Final`::
+
+    class Foo:
+        number: Final[int]
+
+        def __init__(self, number: int) -> None:
+            self.number = number
+
+
+    class Bar:
+        def __init__(self, number: int) -> None:
+            self.number: Final = number
+
+  - Supported by :mod:`dataclasses` (and type checkers since `typing#1669 <https://github.com/python/typing/pull/1669>`_).
+  - Overriding ``number`` is not possible - the specification of ``Final``
+    imposes the name cannot be overridden in subclasses.
+
+* read-only proxy via ``@property``::
+
+    class Foo:
+        _number: int
+
+        def __init__(self, number: int) -> None:
+            self._number = number
+
+        @property
+        def number(self) -> int:
+            return self._number
+
+  - Overriding ``number`` is possible. *Type checkers disagree about the specific rules*. [#overriding_property]_
+  - Read-only at runtime. [#runtime]_
+  - Requires extra boilerplate.
+  - Supported by :mod:`dataclasses`, but does not compose well - the synthesized
+    ``__init__`` and ``__repr__`` will use ``_number`` as the parameter/attribute name.
+
+* using a "freezing" mechanism, such as :func:`dataclasses.dataclass` or :class:`typing.NamedTuple`::
+
+    @dataclass(frozen=True)
+    class Foo:
+        number: int
+
+
+    class Bar(NamedTuple):
+        number: int
+
+  - Overriding ``number`` is possible in the ``@dataclass`` case.
+  - Read-only at runtime. [#runtime]_
+  - No per-attribute control - these mechanisms apply to the whole class.
+  - Frozen dataclasses incur some runtime overhead.
+  - ``NamedTuple`` is still a ``tuple``. Most classes do not need to inherit
+    indexing, iteration, or concatenation.
+
+.. _protocols:
+
+Protocols
+---------
+
+Paraphrasing `this post <https://github.com/python/typing/discussions/1525>`_,
+there is no way of defining a :class:`~typing.Protocol`, such that the only
+requirements to satisfy are:
+
+1. ``hasattr(obj, name)``
+2. ``isinstance(obj.name, T)`` [#invalid_typevar]_
+
+The above are satisfiable at runtime by all of the following:
+
+1. an object with an attribute ``name: T``,
+2. a class with a class variable ``name: ClassVar[T]``, [#invalid_typevar]_
+3. an instance of the class above,
+4. an object with a ``@property`` ``def name(self) -> T``,
+5. an object with a custom descriptor, such as :func:`functools.cached_property`.
+
+Note that the attribute being marked ``Final`` or the property defining a setter
+do not impact this.
+
+The most common practice is to define such a protocol with a ``@property``::
+
+    class HasName(Protocol):
+        @property
+        def name(self) -> T: ...
+
+Type checkers special-case this definition, such that objects with plain attributes
+are assignable to the type. However, instances with class variables and descriptors
+other than ``property`` are rejected.
```

Comment:
Is this consistently implemented? It doesn't appear to be specified in detail, though [the spec](https://typing.readthedocs.io/en/latest/spec/protocol.html) says "To define a read-only protocol variable, one can use an (abstract) property."

Reply:
Huh. I have been relying on what pyright told me the whole time; turns out this is not true.
[mypy](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=12d556bb6ef4a9a49ff4ed4776604750), [pyre](https://pyre-check.org/play?input=%23%20pyre-strict%0Afrom%20abc%20import%20abstractmethod%0Afrom%20functools%20import%20cached_property%0Afrom%20typing%20import%20ClassVar%2C%20Protocol%2C%20final%0A%0A%0Aclass%20HasFoo(Protocol)%3A%0A%20%20%20%20%40property%0A%20%20%20%20%40abstractmethod%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20...%0A%0A%0A%23%20assignability%0A%0A%0Aclass%20FooAttribute%3A%0A%20%20%20%20foo%3A%20int%0A%0Aclass%20FooProperty%3A%0A%20%20%20%20%40property%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20return%200%0A%0Aclass%20FooClassVar%3A%0A%20%20%20%20foo%3A%20ClassVar%5Bint%5D%20%3D%200%0A%0Aclass%20FooDescriptor%3A%0A%20%20%20%20%40cached_property%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20return%200%0A%0Aclass%20FooPropertyCovariant%3A%0A%20%20%20%20%40property%0A%20%20%20%20def%20foo(self)%20-%3E%20bool%3A%20return%20False%0A%0Aclass%20FooInvalid%3A%0A%20%20%20%20foo%3A%20str%0A%0Aclass%20NoFoo%3A%0A%20%20%20%20bar%3A%20str%0A%0A%0Aobj%3A%20HasFoo%0Aobj%20%3D%20FooAttribute()%20%20%23%20ok%0Aobj%20%3D%20FooProperty()%20%20%20%23%20ok%0Aobj%20%3D%20FooClassVar%20%20%20%20%20%23%20ok%0Aobj%20%3D%20FooClassVar()%20%20%20%23%20ok%0Aobj%20%3D%20FooDescriptor()%20%23%20ok%0Aobj%20%3D%20FooPropertyCovariant()%20%23%20ok%0Aobj%20%3D%20FooInvalid()%20%20%20%20%23%20err%0Aobj%20%3D%20NoFoo()%20%20%20%20%20%20%20%20%20%23%20err%0Aobj%20%3D%20None%20%20%20%20%20%20%20%20%20%20%20%20%23%20err%0A%0A%0A%23%20explicit%20impl%0A%0A%0Aclass%20FooAttributeImpl(HasFoo)%3A%0A%20%20%20%20foo%3A%20int%0A%0Aclass%20FooPropertyImpl(HasFoo)%3A%0A%20%20%20%20%40property%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20return%200%0A%0Aclass%20FooClassVarImpl(HasFoo)%3A%0A%20%20%20%20foo%3A%20ClassVar%5Bint%5D%20%3D%200%0A%0Aclass%20FooDescriptorImpl(HasFoo)%3A%0A%20%20%20%20%40cached_property%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20return%200%0A%0Aclass%20FooPropertyCovariantImpl(HasFoo)%3A%0A%20%20%20%20%40property%0A%20%20%20%20def%20foo(self)%20-%3E%20bool%3A%20return%20False%0A%0Aclass%20FooInvalidImpl(HasFoo)%3A%0A%20%20%20%20foo%3A%20str%0A%0A%40final%0Aclass%20NoFooImpl(HasFoo)%3A%0A%20%20%20%20bar%3A%20str%0A) both allow `ClassVar`s and descriptors, only [pyright](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBAhgRgYygSwgBzCALrOBnLEGBLCAUywAswATAKFEimAFcA7EsMAGzxXUw4ExSmRoB9NODRlsATwbhoWOWmRsA5vwzYoAYW4w8eAGowQAGigAFcFjAIeV4Opjc6HhIeNQAEkYAxLgAKWzB7R24ASgAuOigEqAABKTAZeXjEpPgCIhJyKlpMhJoyYGYQvDJuYCioAFoAPhQ2LBioADoujzoAYlhjZA02eGRuZBUeryM%2BILAAQSxCZDgWLDI4xIqwdvUsT29ZrjD0lU2s1NOFLdLy4Erq2obmvfaQChYQNigABgOZqBzAwzMwgc4Je47fSHUEAbT2AF0oABeX7-HxzAAiZDwCBAyDQ9jBxWSwgQogkl1kkxuZW2wSqNTqTRabSg7ywn2%2Bfzo0wxx2k1LkejAADdzMgYK1wckqRlaXcHkznlA4FxuG8Pl9AW4quijmAAJJscXjGgyyHtXL6qAAOTAcxlcHMVsIPTAcAAVu1-Hg5nQPZ6UYCuItlqt1sE6lB%2BmAANYBr3BuYnIVRxKxhOB5NcYHGUFbGNQeOJoOooEw8zphKZ0s5sDY3H4wmYdO17PlgVpIUi8X4qVYNvFrNJztGk1uZA0as1qCyEB11H2uYzwtF%2BeLu1gNhkNdr-objwHgAeaHGCAm2ncNrmYfxEbIhvQ3GCvrmsRJltZN67VyfZ9fQIuA-LYUkFeVEluelGSeFlXnZLVuR-MA81Mcx-xfN9gItLh2lQuFEWDHk%2BQNRs8QJIkMMAv1sJJJIyQpSRwJpSC6UhBlHmZF5pQQzltWIw4QzAVN5F7CUByorCwBAi5mOuVjFTADjlRZNUeE1PjvgCXUyGQ41TSnSSgOknCoWtOgkhcEZ3BIrc5iMmiTJJZ0wSga0gA) rejects them.
They also support explicitly declaring the implementation via inheritance.

This changes the whole narrative, which is troubling. It would seem that only a minor update to the spec is required so as to have all type checkers consistent here